python/src/search_algorithm.py

```diff
@@ -46,35 +39,47 @@ def get_next_population(
 class SearchAlgorithm(Generic[Candidate, SimulationResult]):
     __metaclass__ = ABCMeta
 
-    cost_function: CostFunction
     candidate_generator: CandidateGenerator[Candidate]
+    _cost_function: Optional[CostFunction]
+    _simulate: Optional[Simulator]
+    candidates: Sequence[Candidate]
+    iteration: int
+    simulation_result: SimulationResult
+    cost_map: CostMap
```

Comment:
Not a big fan of having mutable state... Any reason why these can't just be local vars to the `search` method? If the plan is to have them visible outside the class maybe there should be a `SearchResult` that returns those things?

Reply:
I made them members (ideally they would be private but I don't think you can enforce that in python), so that they could be accessed by the convergence detector, i.e. `should_stop`

Reply:
If you prefix with `_` i think pylint will enforce it. Yeah I figured the algorithm that implemented `SearchAlgorithm` would decide what members to make available to `_should_stop`, but i guess there's no harm by exposing all by default